Added support for USR version 5 and 6 and corrected USR version 4 support.

bjkowalski opened the pull request:
Eliminated separate processing for USR4 to simplify user interface.  Now processing detects the USR version and does the appropriate processing.  Newer Lowrance units generate different USR2 than what was supported.  Corrected old USR4 processing and added support for USR5 and USR6. 

Added tests for USR4, USR5, and USR6 and the ability to compare results between them in some cases.
Renamed test files to follow a lowrance-*.* naming scheme.  Significant documentation updates.

bjkowalski committed:
Minor changes to address GCC 7.3.0 warnings generated by flag -Wunused-but-set-variable.

Moved declaraions/initialization of rp_backup and era inside #ifdef _NL_CURRENT conditionals.

bjkowalski committed:
Added { ... } to for at the end of zip64local_SearchCentralDir() so that if-break statements inside for loop.

bjkowalski committed:
Removed declaration and commented out references to startlat, startlon, and starttime in fit_parse_data() due to GCC 7.3.0 warning from
flag -Wunused-but-set-variable.

bjkowalski committed:
Added initial support for Lowrance USR 5 and 6 formats.

bjkowalski committed:
Significant changes to handle multiple USR4 versions and add USR5 and USR6 support.

bjkowalski committed:
Remove extraneous debug NL when no debug specified.

bjkowalski committed:
Updated USR4 reference file because of increased accuracy in converting lat/log values.

bjkowalski committed:
Documentation on some ICON mappings.

bjkowalski committed:
Added USR5 and USR6 to validation checks.

bjkowalski committed:
Added reference files for USR5 and USR6 formats.

bjkowalski committed:
Cleaned up some debug output.

bjkowalski committed:
Continued work to consolidate all Lowrance USR processing into a single file.

bjkowalski committed:
Eliminated lowranceusr4.cc and restuctured testing to include USR4, USR5, and USR6 formats.

bjkowalski committed:
Changes associated with adding support for USR 5 and USR 6.

bjkowalski committed:
Eliminated lowranceusr4.cc from build.

bjkowalski committed:
Changed default Output USR Version from 4 to 2 for backward compatibilty.

bjkowalski committed:
Removed USR 4 section from documentation.

bjkowalski committed:
Initial update of XMLDOC for updated data formats.

bjkowalski committed:
iMinor updates to parse_routes.  Added correct v4 and v5 GPX reference files.

bjkowalski committed:
Documentation updates to reflect new capabilities.

bjkowalski committed:
Added Route element information to document and minor code changes to match.

bjkowalski committed:
Doc refinement with rewrite.

bjkowalski committed:
Corrections to USR 4 output.

bjkowalski committed:
Update to correct v4 gpx file.

bjkowalski committed:
Checkin correct v5 gpx file.

bjkowalski committed:
Removed errant NL when no debug specified.

bjkowalski committed:
Resolved conflicts

bjkowalski committed:
Merge branch 'master' into master

bjkowalski committed:
Missed this update.

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

tsteven4 commented on lowranceusr.cc:
For maintenance it is very nice to have the sources of information available.  Can you list the URLs?

bjkowalski replied:
Added URLs per request.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -489,167 +1230,320 @@ lowranceusr_parse_icons()
 
 static void
-lowranceusr_parse_trails()
+lowranceusr_parse_trail(int *trail_num)
```
There appears to be no intent to modify the callers value of trail num.  Pass as int?

tsteven4 replied:
my bad, break trail into segments modifies trail_num.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -489,167 +1230,320 @@ lowranceusr_parse_icons()
+}
+
+static void
+lowranceusr4_parse_trail(int *trail_num)
```
There appears to be no intent to modify the callers value of trail num. Pass as int?

bjkowalski replied:
Left this in for consistency with lowranceusr_parse_trail().  Like lowranceusr_parse_trail() this may need to update if I find that trails need to be broken into segments.

tsteven4 commented on lowranceusr.cc:
copyright issues?

bjkowalski replied:
I wondered about that as well, hence my reference to the source.  This is also available electronically and will include the URL per your earlier suggestion.  I've included similar text in the doc updates.

tsteven4 replied:
Please remove any and all text from copyrighted sources that we don't have permission to use.  This applies to all files, including source and documentation files.  You may substitute a phrase like:
"The recipe for green eggs and ham may be found in the book "Green Eggs and Ham" by Dr. Seuss, which is available at
http://www.seussville.com/books/book_detail.php?isbn=9780394800165

In the document you can use ulink elements, for an example see xmldoc/formats/kml.xml

bjkowalski replied:
Replaced copyrighted material with authored comments.

tsteven4 commented on testo.d/lowranceusr.test:
```diff
@@ -1,36 +1,150 @@
+#
+# File Name		  	File Contents
+# ===============		========================
+# lowrance.usr			USR version 2 format file.  Generated by GPSBabel from geocaching.loc input..
```
Its very nice to document where reference files came from, and which are golden, i.e. directly from the device.  They way we store reference files and translations of reference files in the same directory is confusing at best.

tsteven4 commented on testo.d/lowranceusr.test:
```diff
@@ -1,36 +1,150 @@
+
+# Should have generated the same GPX file as USR version 4 but the order is different for some reason
+# so use another way to compare to ensure have EXACTLY the same data
+#
```
Interesting.  Maybe we shouldn't be storing rte_head->rte_num values at all if the order we retrieve the routes/trails doesn't mean anything.

If didn't store rte_num values then I have enhancements to the sort filter that could resort the routes & tracks by name, description or number.  The sort filter can already sort the waypoints by time, etc.

bjkowalski replied:
Being able to sort would definitely simplify the verification tasks, and I could see where there would be advantages when moving from one vendor to another.  Right now there is so much loss of Waypoint, Trail and Route information moving from USR to GPX, or even USRx to USRy, that the change would seem to only benefit the validation process.  I'd definitely support adding the option but without significant support from Lowrance on decoding the USR formats it wouldn't be something that many folks would use at least for sorting data (which is not an option they have on their units).

tsteven4 replied:
I merged #273.  I think it is of general usefulness, and if you want to use it for verification of lowranceusr it's available.  However, what you have works fine, and there is not need to change it.
For example
gpsbabel -i lowrancusr -f ${REFRENCE}/lowrance-v5.usr -x sort,rtename -o gpx -F lowrance-v5.gpx

However, because the route number (rte_num) values reflect the order in the usr file, while the route names don't, the gpx files still won't match between usr versions.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -489,167 +1230,320 @@ lowranceusr_parse_icons()
+    double longitude   = lon_mm_to_deg(gbfgetint32(file_in));
+    int    icon_number = gbfgetint32(file_in);
+
+    if (opt_ignoreicons == NULL) {          /* Option not specified if NULL */
```
nullptr instead of NULL.

bjkowalski replied:
Changed all NULL references to nullptr.
Thanks.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -894,18 +1884,75 @@ lowranceusr_route_hdr(const route_head* rte)
+  /* UID unit number */
+  if (opt_serialnum_i > 0)
+    gbfputint32(opt_serialnum_i, file_out);   // use option serial number if specified
+  else if (rte->fs != (format_specific_data *)NULL)
```
nullptr instead of "(format_specific_data *) NULL" ?

bjkowalski replied:
Adopted change.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -758,24 +1652,120 @@ lowranceusr_waypt_disp(const Waypoint* wpt)
+  /* UID unit number */
+  if (opt_serialnum_i > 0)
+    gbfputint32(opt_serialnum_i, file_out);   // use option serial number if specified
+  else if (wpt->fs != (format_specific_data *)NULL)
```
nullptr instead of "(format_specific_data *) NULL" ?

bjkowalski replied:
Adopted suggestion.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -894,18 +1884,75 @@ lowranceusr_route_hdr(const route_head* rte)
+
+
+static void
+lowranceusr4_route_trl(const route_head* rte)
```
"const route_head* rte" -> "const route_head* /* rte */" or "const route head*" will silence warnings about unused parameter.

bjkowalski replied:
Thanks for the info.

bjkowalski committed:
Revert "Minor changes to address GCC 7.3.0 warnings generated by flag -Wunused-but-set-variable."
at the request of tsteven4 as they have nothing to do with the expanded support.

This reverts commit fb93a19b740edae4dde3b6dcd0546f344122f70d.

bjkowalski committed:
Revert "Added { ... } to for at the end of zip64local_SearchCentralDir() so that if-break statements inside for loop."
at the request of tsteven4 as they have nothing to do with the expanded support.

This reverts commit 189bbe6ea664829412c70079e9ddb9555062a12f.

bjkowalski committed:
Revert "Removed declaration and commented out references to startlat, startlon, and starttime in fit_parse_data() due to GCC 7.3.0 warning from"
at the request of tsteven4 as they have nothing to do with the expanded support.

This reverts commit cc618336888c5cdb8adef87f72026a8c24eff7e8.

bjkowalski committed:
Updated to remove lowranceusr4.cc per suggestion from tsteven4 to support cmake/XCLion.

bjkowalski committed:
Per suggestions from tsteven4
    - deleted define SECSTORADIANS as it was not used
    - ran astylerc

bjkowalski committed:
Replaced use of NULL with nullptr per tsteven4 suggestion.

bjkowalski committed:
Added URLs for reference information.

bjkowalski committed:
Adopted tsteven4 recommendation on eliminating compiler warning.

bjkowalski committed:
Corrections to Waypoint and Trail documentation for USR 4, 5, and 6 formats.

bjkowalski committed:
Address Codacy flagged issues.

bjkowalski committed:
Address Codacy flagged issues.

tsteven4 commented on lowranceusr.cc:
I would still like to see this:
#define DEGREESTORADIANS (M_PI/180.0)
If not defined elsewhere defs.h will supply M_PI.
The compiler should do the division at compile time.
This will reduce the chance of reference file thrashing in the future as there isn't the temptation to add another digit.

bjkowalski replied:
I was planning to make the change but I need to be able to spend enough time so I can make the change, regenerate the reference files, and then desk check to make sure I didn't mess anything up.  I am hoping to get that done in the next few days.

bjkowalski replied:
Suggested changed adopted.

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

bjkowalski committed:
Merge branch 'master' of https://github.com/bjkowalski/gpsbabel

bjkowalski committed:
Address additional Codact flaged issues.

bjkowalski committed:
Changed DEGREESTORADIANS define from hardcoded value to use (M_PI/180.0)

bjkowalski committed:
Removed Copyright protected material and replaced with authored comments.

bjkowalski committed:
Corrected altitude/depth retension coverting USR4/5/6 to USR2/3

bjkowalski committed:
Added mapping of USR4/5/6 ICON numbers to text to support output to USR 2/3 format.

bjkowalski committed:
Corrected generation of USR4 data file.

bjkowalski committed:
Check for NULL format specific when converting USR 2/3 (colorless) to USR4 (color coded) ICONs.

bjkowalski committed:
Added cases where USR 2/3 embeds color in shortname.

bjkowalski committed:
Corrections to ICON translation from USR v2/3 to USR 4.

bjkowalski committed:
Cleanup USR 2/3 parsing debug.

bjkowalski committed:
Adopted modificatins to USR 4/5/6 time interpretation per suggestion from tsteven4.

bjkowalski committed:
Adopted modificatins to USR 4/5/6 time interpretation per suggestion from tsteven4.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -423,28 +960,397 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
-  if (reading_version >= 3) {
+  if (reading_version == 3) {
     float depth_feet = gbfgetflt(file_in);
     if (std::abs(depth_feet - 99999.0)  > .1) {
```
L973 implies the field is stored as a float.
L974 implies that the magic value +99999.0 indicates no data.
compare this with the comment at L1808-1809.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -758,24 +1802,140 @@ lowranceusr_waypt_disp(const Waypoint* wpt)
+  if (writing_version == 3) {
+    float depth = WAYPT_HAS(wpt, depth) ?
+                  METERS_TO_FEET(wpt->depth) : -99999.0;
+    gbfputint32(depth, file_out);
```
L1808 implies the magic value -99999.0 indicates no data.
L1809 converts the float depth to an int, and saves it as an int, not a float.
Compare to comment at L973-974.
These operations should be symmetric between read and write.

tsteven4 commented on lowranceusr.cc:
```diff
@@ -423,28 +960,397 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+  char unused_byte = gbfgetc(file_in);
+
+  /* Altitude/Depth in feet */
+  wpt_tmp->altitude = FEET_TO_METERS(gbfgetflt(file_in));
```
I am not sure altitude is what you want.  I see lots of ele elements now with values of 0.0 in gpx output generated from usr input.  This suggests that 0.0 values are being falsely interpreted as valid sea level.  Perhaps there is a flag somewhere in the usr file that indicates if this field is valid (like in the unused byte just before this?).
Internally we use the magic value unknown_alt to indicate invalid altitude data in a waypoint. We are not consistent in how we handle unknown values (magic values like unknown_alt vs. flags like wpt_flags bit vs. gpsbabel_optional::optional.)

The definition of depth in defs.h regarding "thickness" of a waypoint may not be used consistently.

What was the case you mentioned with issues 200ft below sea level?  Is that related to this change?

In any event read and write need to be symmetric, here you currently read into altitude, but in lowranceusr4_waypt_disp you write from depth.

tsteven4 replied:
2016 baja 1000 has usr versions 2, 4 and gpx available.  A few of the points in the prerun gpx file have non-zero elevations, but their gpx has mostly ele elements with 0 values which are likely incorrect (the element is optional and shouldn't be included if the data is unknown).  From the position we should be able to confirm the elevations from DEMs.

tsteven4 replied:
I think usr v4 elevation is stored in what we call Flags.  I suspect what you now call Altitude/Depth is something else, perhaps depth.  We think usr v3 has both altitude and depth.

  /* Long/Lat */
  wpt_tmp->longitude = lon_mm_to_deg(gbfgetint32(file_in));
  wpt_tmp->latitude = lat_mm_to_deg(gbfgetint32(file_in));

  **/* Flags, discard for now */
  fsdata->flags = gbfgetint32(file_in);**

tsteven4 replied:
Scratch that, its not in Flags.  somewhere else.

bjkowalski replied:
I was able to find a USR2 file on an iFinder HUNTc which had waypoints with lat, lon, elev, date, time values.  With the updates now committed the output using unicsv generates the same values as displayed on the iFinder device.  In that process I discovered another way Lowrance indicates unknown elevation in USR2 files.  There are now two different check methods when reading elev in USR2 and USR3 files.

bjkowalski replied:
Issues with inconsistent use of altitude and depth have been addressed.

bjkowalski committed:
Changed file header parsing back to int16 for USR format as found sample file with a stream version other than 0.  Assume this
is true for ALL USR formats until determine otherwise.

bjkowalski committed:
Updated reference files after corrections to altitude and time conversion.

bjkowalski committed:
resolved conflicts in lowranceusr.cc

GPSBabelDeveloper commented on lowranceusr.cc:
Nobody's noticed in all these years, but is this perhaps a "Letterbox cache"?

Described at https://www.geocaching.com/about/cache_types.aspx

GPSBabelDeveloper commented on lowranceusr.cc:
Throughout you use gree*m*. Did you mean "green"? If all 7 elements of this really are the same everywhere, we should reconsider a more concise data structure...

bjkowalski replied:
They should all be green.

While there are only 7 values (0-6) used to represent colors, looking only at what the unit displays I couldn't derive a method for consistently representing what was being displayed without separate, explicit values.  Some like anchor and rock appear to be the same icon displayed but the unit allows me to pick from seven icons and each one has a different color value.  It also appears to me, from looking at the unit, that there are actually 8 colors (blue, magenta, orange, yellow, green, aqua, white, red) represented by the 7 values (0-6).

I need access to a higher quality screen than the 7" 800x480 display on my unit.

GPSBabelDeveloper commented on lowranceusr.cc:
Should waypt_table (and _ct and _sz) be something like a QVector<Waypoint*> or maybe a QList? I don't this is really a high-running path, but these linear walks of the structure and the manual resizing below make me cringe.
This CL is already large. If we defer this to a future point, so be it.

GPSBabelDeveloper commented on lowranceusr.cc:
throughout this patch, there's a lot of struggle between C and C++/Qt. (I know this didn't start this week. I'm just talking from bed...)
qDebug() << MYNAME << "adding waypt" << wpt->shortname<<  "to index" << waypt_table_ct;
At some point, I plan to provide a custom class that unifies our formatting, debugging levels(byte level, logical) but it's not here yet. This patch just spends a LOT of effort formatting deb code and at some point, we should ask if we're doing the debugger's work...

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -248,12 +547,93 @@ lowranceusr_readstr(char* buf, const int maxlen, gbfile* file)
+
+//  Starting with USR 4, adopted a UTF-16 character string format
+static int
+lowranceusr4_readstr(char* buf, const int maxlen, gbfile* file, int bytes_per_char)
```
Bytes per_character is always either 1 or 2, right? It's either ASCII or UTF-16, right?
Can this whole function become something like:
QString lowrance4_readstr(gbfile, int bytes_per_char)  {
int org = len = gbfgetint32(file);
buf = xmalloc(len) // Probably something more clever here to be done
bytesread = gbfread(buf, bpc, len, file);
QString rval;
if (bpc == 1) {
  rval = QString::fromUtf8(buf, len);
} else {
  rval = QString::FromUtf16(buf, len)
}
return rval;

Maybe we give up on gbfread (word size, endian safety, ...) and make everything QFiles and then we can stream QByteArrays directly.

tsteven4 replied:
I already have a patch that uses codecs for usr 4 readstr and writestr.

bjkowalski replied:
Adapted the codec patch.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -248,12 +547,93 @@ lowranceusr_readstr(char* buf, const int maxlen, gbfile* file)
+  if (bytes_per_char == 1) {
+    (void) gbfwrite(CSTR(buf), 1, len, file);
+  } else {
+    for (int i = 0; i < len; ++i) {
```
Maybe writestr() also gets a similar toUtf() and toUcs16() beating.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -355,34 +842,63 @@ lon_mm_to_deg(double x)
+   */
+  if (rstream_version == 0) {
+    float read_alt = gbfgetflt(file_in);
+    char buf[16];
```
this hunk is weird. Using sprintf and strcmp to determine isnan isn't great. Is this basically
read_alt = getflt() (may rely on safe upcast of float todouble)
if isNaN(read_alt)) {
wpt_tmp->altitude = unknown_alt; (Steve, did we just not get optional to alt because of the zillion places it's used?)
} else {
  wpt_tmp->altitude = FTOMeters()
} else {
 wpt_tmp->altitude = ...
}

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -391,9 +907,25 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
-           wpt_tmp->longitude, wpt_tmp->altitude);
+  if (global_opts.debug_level > 1) {
+    if (global_opts.debug_level == 99) {
+      printf(" %16.16s %+15.10f %+15.10f",
```
This is an awful lot of finicky formatting in debug code. Are we doing the debuggers's job for it? Could the finicky interleaving between string constants and debug values just be done with qDebug() ? Note http://doc.qt.io/qt-5/qdebug.html#formatting-options if you really want to display tenths of feet.
Maybe more readable to bust all this type of code into different functions and just pass the pointer to the wpt around.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -406,15 +938,35 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+      printf(" %s %s", qPrintable(dt.toString("yyyy/MM/dd")), qPrintable(t.toString("hh:mm")));
+    } else {
+      printf(MYNAME " parse_waypt: creation time %d, base_time %d waypt_time %d\n",
+             (int)wpt_tmp->creation_time.toTime_t(),
```
Notably qDebug knows how to print a QDateTime in a way that's human-readable. qDebug() << "created" << wpt_tmp->creation_time << "base time" << base_time_secs should do something reasonable.
It's not really suited to the writing of the files, but it saves that ickiness in both of these two printf lines.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -406,15 +938,35 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+  wpt_tmp->icon_descr = lowranceusr_find_desc_from_icon_number(icon_number);
   if (wpt_tmp->icon_descr.isNull()) {
     char nbuf[10];
     snprintf(nbuf, sizeof(nbuf), "%d", le_read32(buff));
```
wpt_tmp->icon_descr = QString::number(le_read(buff));

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+static void
+lowranceusr4_parse_waypt(Waypoint* wpt_tmp)
+{
+  char name_buff[MAXUSRSTRINGSIZE + 1];
```
Put these closer to their actual use. Top-of-function quit being a rule long ago. (Yes, a lot of the code in GPSBabel is really old...)

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+  fsdata->color_desc = lowranceusr4_find_color_from_icon_number_plus_color_index(fsdata->icon_num, fsdata->color);
+
+  /* Waypoint descr; input is 2 bytes per char, we convert to 1 */
+  int desc_len = lowranceusr4_readstr(&desc_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
```
The repeated blocks of boilerplate are a good indication that this calling convention needs "fixed" to return  useful value.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+      printf(" %6d %16s", desc_len, desc_buff);
+      QDateTime dt;
+      QTime t;
+      dt = wpt_tmp->GetCreationTime();
```
Changing all the printfs to qDebugs gets the variable creation and manipulation out of this nested if debug/if more debuglevel.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -423,28 +975,399 @@ lowranceusr_parse_waypt(Waypoint* wpt_tmp)
+      if (reading_version > 4) {
+        printf("-------- -------- -------- -------- ");
+      }
+      printf("----------- -------- -------- -------- ------ ----------------");
```
This will admittedly be difficult/fragile with qDebug(), but it's probably fragile here, too. Maybe it all becomes a qtextstream and http://doc.qt.io/qt-5/qtextstream.html#setFieldWidth and let that handle the columnization.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -455,226 +1378,361 @@ lowranceusr_parse_routes()
+      /* position coord lat & long */
+      wpt_tmp->latitude = latitude;
+      wpt_tmp->longitude = longitude;
+      wpt_tmp->altitude = 0;
```
Don't explicitly set alt if we don't know it. (That'll be more important later.(

bjkowalski replied:
Historic code.  I changed it to be "unknown_alt" as that seems more appropriate given that we are converting an Event Marker (icon/lat/lon only) to a waypoint.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -455,226 +1378,361 @@ lowranceusr_parse_routes()
+      wpt_tmp->latitude = latitude;
+      wpt_tmp->longitude = longitude;
+      wpt_tmp->altitude = 0;
+      snprintf(buff, sizeof(buff), "Event Marker %d", i+1);
```
...shortname = "Event marker" + QString::number (i+1);

In general, avoid s[n]sprintf when we can.

bjkowalski replied:
Adopted in the latest QString changes.

GPSBabelDeveloper commented on lowranceusr.cc:
```diff
@@ -455,226 +1378,361 @@ lowranceusr_parse_routes()
+  int desc_len = lowranceusr4_readstr(&desc_buff[0], MAXUSRSTRINGSIZE, file_in, 2);
+  if (desc_len) {
+    desc_buff[desc_len] = '\0';
+    trk_head->rte_desc = desc_buff;
```
WIth foo_readstr() returning QStrings, we'd be out of the business of manually terminating and testing them before we used them.About every caller would benefit.

tsteven4 replied:
I already have a patch that does this.

bjkowalski replied:
Consolidated to a single lowranceusr_readstr() using QStrings.

GPSBabelDeveloper reviewed (commented):
I had to stop, but it seemed I had the biggest part reviewed. Well, that and it's 3am.

Looks good. Most of my comments can to in a future commit. This format is approaching .gdb for complexity and unknown fields. :-/

bjkowalski committed:
Check for both "nan" and "-nan" to address compiler differences.

bjkowalski committed:
Use -options instead of --options to avoid osx OS issues.

bjkowalski committed:
Set altitude to "unknown_alt" when creating Waypoints from Event ICONs.
Updated reference files to encorporate "unknown_alt" change.
Corrected mis-spelling of "green"

bjkowalski committed:
Change cut options from --compliment to legacy explict field list for OSX OS support.

bjkowalski committed:
Applied codex patch from tsteven4 - thank you.

bjkowalski committed:
Corrected interpretation of elevation in USR2/3.

bjkowalski committed:
Corrected reference files and test script with changes to elevation interpretation.

bjkowalski committed:
More conversions to QString conventions.
Minor correction to trail attribute parsing.

bjkowalski committed:
Editorial cleanup.

bjkowalski committed:
USRv6 uses trail data stream version 5 which employs int32 rather than char for attributes.

bjkowalski committed:
Use invalid ICON value as method to detect HOOK2 generated USR 2/3 files.

bjkowalski committed:
Cleanup time conversions.
